### ISSUE TITLE
New Block: Dialog Popup

### ISSUE THREAD

**Issue by bhubbard** [WordPress/gutenberg#61297]
## What problem does this address?
Popups are very common for websites, and its a common request.

## What is your proposed solution?
Creating a new block for popups using the dialog html element: 
https://www.youtube.com/watch?v=ywtkJkxJsdg

**Comment by paaljoachim** [comment]
I am thinking this plugin is something to take a closer look at:
https://wordpress.org/plugins/makeiteasy-popup/

I found out about the plugin as the author mentioned it in the Advanced WordPress Facebook group here:
https://www.facebook.com/groups/advancedwp/posts/8230806923648116/

Here are some of the things the author wrote:
"It is about popup. I developed a popup block and tried to make it extensible and configurable, ready to use in production. It has many features and follows best practices of core blocks development, which means you can style it inside block editor with native block editor controls and you can insert any number of other blocks in it.
I think it is ready for production sites, check it out! Please write your remarks and suggestions, so I can improve it.
I believe block with so many features can easily be charged for, but I offer it for free."

"...to provide a block which would be close to core blocks and give users options which they often need. I started this block when dialog element was very new, I am planning to switch to dialog element."

**Comment by sethrubenstein** [comment]
I'm finalizing a set of `<dialog/>` blocks that utilizes the Interactivity API for a project we're working on for release in February 2025. I could easily modify this to be more Gutenberg/Core friendly, if there's agreement that this would be a good fit for the core block library?

https://github.com/user-attachments/assets/6e57d6d1-ee13-4f67-9872-a580966b8d59

**Comment by paaljoachim** [comment]
I will ping a few folks so they can give feedback.
@SantosGuillamot @michalczaplinski   @WordPress/interactivity-api

**Comment by michalczaplinski** [comment]
I would love to see more Core blocks that leverage the Interactivity API! 🙂 

That said, with dialogs there are a11y concerns that I'm not an expert on. I don't have a strong opinion on whether this should be in Core or is more of a plugin territory but perhaps @WordPress/gutenberg-core does.

**Comment by draganescu** [comment]
Would this offer a good enough experience to be preffered to full blown modals?

**Comment by luisherranz** [comment]
I think this is more for @mtias or @jasmussen to decide, but I agree that exploring a dialog/pop-up block in Core makes sense.

> Would this offer a good enough experience to be preffered to full blown modals?

What do you mean by _full blown modals_?

**Comment by draganescu** [comment]
> What do you mean by full blown modals?

I mean modal windwos and dialogs are provided by plugins generally and they generally use libraries with specifics implemented. I wonder how much we need to pad the support for the html element to make it useful and if we support all the stuff that make it work right.

**Comment by sethrubenstein** [comment]
Hi all, I'm putting some final touches on this vis-a-vis accessibility on the "dialog trigger" block this week and I'm hoping to submit a PR to start a more in depth review discussion early next week. 

I've tried to add as much support for the native `<dialog/>` element as possible, including color controls for the `::backdrop` pseudo element, the ability to toggle between the native `modal` and `dialog` types. There are some other nice to haves as well like an auto-activation timer. 

@draganescu Regarding support for other plugins, absolutely, that's been a key consideration in our design. Our first major use of this block will actually be programmatically generated dialog blocks in an upcoming project. So, getting extensibility right was key for even our own use-case. As for other plugins interacting with the dialog block(s): 

For starters open and closing a dialog is as easy as `store('core/dialog').open(theIdOfTheDialog)`, as well as `store('core/dialog').close(theIdOfTheDialog)`, `store('core/dialog').closeAll()`. 

We also have some additional functionality that will play or pause any VideoPress video inside a dialog when opening and closing. I'm going to pull that out before submitting the PR, but I'll provide a code example for how a video player block can easily opt-in to the dialog block's interactivity store to support an open/play close/pause interaction.

**Comment by djcowan** [comment]
Personal thoughts; `dialog` windows by tradition prompt users for text input (forms), whereas `modal` windows are used to alert or inform users to system changes. The distinct typs of modal components evades me at present but non-draggable 'confirmation' and dismissible notifications fell into that category.

Aside from registration || subscription forms, I'm failing to see how a `modal/dialog` component would fit into the Wordpress ecosystem.

Perhaps I am simply being a grumpy old so-and-so but there was a reason why pop-up blockers were created

**Comment by sethrubenstein** [comment]
I personally have found dialog windows to be more useful for just prompt user's to input data or alert them. We use dialogs generally to display a variety of content above other content. This certainly may be not a good fit for the core block library (once I get around to submitting it), that said, I personally believe there should at least be a core block equivalent for every html element and as `<dialog/>` is a native element now it's worth exploring.

**Comment by Mamaduka** [comment]
It could be a good candidate for #58773.

**Comment by djcowan** [comment]
> It could be a good candidate for #58773.

Excellent proposal.

**Comment by TheJeffr0** [comment]
How similar would this new block be to what is described here? https://designsystem.digital.gov/components/modal/

**Comment by sethrubenstein** [comment]
@thejeffro Very similar in functionality but this uses the native browser `<dialog/>` element. This will let you do modals yes, but also "popover" style interface elements as well.

**Comment by TheJeffr0** [comment]
If that's the case, I'm on board with this new block. Would love to see it implemented.

**Comment by felixarntz** [comment]
I'm excited to see this effort move forward, also in regards to relying on the native `dialog` element. Would be happy to review an eventual PR @sethrubenstein!

**Comment by scruffian** [comment]
Great idea!

**Comment by vk17-starlord** [comment]
**Hey @Mamaduka @felixarntz** — would it make sense for me to start working on adding a new dialog popup block to Gutenberg?

If yes, I’d love to get some initial clarity from the core team to ensure I’m aligned with the expectations and use cases. Here are a few questions I have to get started:

1. **What is the intended primary use case** for the dialog popup block? (e.g., modal for calls-to-action, alerts, forms, etc.)
2. **Should it follow existing accessibility guidelines and practices** used in other Gutenberg interactive blocks?
3. **Are there any design mocks, Figma files, or references** available for how this block should look or behave?
4. **Do we want this block to be fully customizable** (positioning, trigger method, animation, etc.), or scoped to specific layouts or behaviors?
5. **Should the block support dynamic content** like block patterns or reusable blocks inside the dialog?
6. **Are there any known limitations or blockers** that I should be aware of before starting?

Happy to start exploring and prototyping once I have a better understanding of what’s expected. Let me know how you'd like to proceed!

**Comment by Mamaduka** [comment]
@vk17-starlord, feel free to work on it, but the new blocks are low priority at the moment. There's also a discussion on how new blocks are shipped in core - https://github.com/WordPress/gutenberg/issues/58773.

**Comment by vk17-starlord** [comment]
Sure @Mamaduka will look how new blocks are shipped in core and will create block accordingly !!

**Comment by bhubbard** [comment]
@vk17-starlord 

1. Yes, this should be independent of the post loop.
2. username, First Name, Last Name, email, gravatar, bio, website, custom meta, should all be on the list
3. Yes, I think it should mimic the Query Loop as much as possible.
4. Probably, don't know enough to fully answer this question.
5. IT can start without custom user meta fields, although I think long term it will need to support it.

**Comment by vk17-starlord** [comment]
# Dialog Popup Block - Input Request

I'm working on the Dialog Popup block and need your input on the user experience.

## Key Questions

### Editor Experience
- How should users edit content inside the popup? 
- Should users be able to drag and drop other blocks inside the popup?
- Should the popup content be visible in the editor, or should we provide a button to toggle it?

### Customization Options
What options are most important to users?
- Dialog size/position
- Colors and styling
- Animation effects
- Close button style
- Trigger options (button, auto-trigger, etc.)
- Behavior settings (dismissible, ESC key close)

### Use Cases
What are your main use cases for a dialog popup?
- Forms
- Image galleries
- Notifications
- Other?

@bhubbard @Mamaduka @t-hamano

**Comment by bhubbard** [comment]
@vk17-starlord 

The block should support drag and drop for other blocks, I would imagine something similar to the details box. I would imagine forms, notifications, and marketing promotions as the most common use cases. The list you provided for customization options is a good start (take out animation for a v1), I would maybe look at some other popular pop-up builders for WordPress, but find the balance to keep the options simple to start.

It might make sense to have the close button be another custom block that only works within the dialog block. This way a user can move it to anywhere they want and style it as needed.

**Comment by sethrubenstein** [comment]
@vk17-starlord You may want to take a look at what we have in prc-block-library as a starting point. I had intended to bring a PR for our dialog implementation but have been focused on getting #69789 ready. The biggest "pain point" in creating a Core Dialog block will be UI/UX and accessibility on actually triggering the dialog, thats where the most thought and time will be needed IMO on this block. 

https://github.com/pewresearch/prc-block-library/tree/main/src/dialog
https://github.com/pewresearch/prc-block-library/tree/main/src/dialog-trigger
https://github.com/pewresearch/prc-block-library/tree/main/src/dialog-element

And see https://www.pewresearch.org/religious-landscape-study/age-distribution/18-29/?activeChartId=6ac4c46314ff76e21be70e1f66fe1a19&dialogId=31585855a95a80d3a7aabf7d2cff9c62 for a programmatic usage. For a user interface element like a Dialog, it'll be important to allow deep linking to activate dialogs.

**Comment by vk17-starlord** [comment]
Thanks @sethrubenstein  for sharing this, really appreciate the context and links! 🙌
I'll go through the implementation in prc-block-library, especially the dialog, trigger, and element structure. The point on UI/UX and accessibility around triggering makes sense — definitely agree that's where most of the polish and thought will be needed.

**Comment by sethrubenstein** [comment]
An update here: I'm hoping to have a PR by Wed with an initial pass at porting dialog, dialog-trigger, and dialog-element blocks out of prc-block-library and into Gutenberg's core block library.

**Comment by sethrubenstein** [comment]
A day late and **some more work to do on animations and the styles and undoubtedly accessibility**, however I wanted to get this submitted so people can start testing and thinking of any improvements we can make. https://github.com/WordPress/gutenberg/pull/71618

## `core/dialog`

`core/dialog` is actually a collection of three blocks: 1. `core/dialog` a wrapper really for 2. `core/dialog-trigger`, and 3. `core/dialog-element`. 

### `core/dialog-trigger`

This block is the "trigger" or "action" to open the dialog-element `<dialog/>` block. 

### `core/dialog-element`

This is the primary block and is a representation of the `<dialog/>` element in modal form. With added support for positioning center, top left, top center, top right, center left, center right, bottom left, center, and right and support for backdrop coloring it utilize core supports for box shadow and border quite well. 

Additionally, this block and it's iAPI store have been designed to be maximally extensible. We utilize this block throughout pewresearch.org, most complexly in RLS https://www.pewresearch.org/religious-landscape-study/age-distribution/18-29/?dialogId=dialog_prayer-frequency&activeChartId=6ac4c46314ff76e21be70e1f66fe1a19. This link shows off another feature built into `core/dialog`, in this case a modified version of it's built in Deep Linking feature. Which allows you to open a dialog on page render so long as it's id is in the url with `?dialogId`. There is also an Auto Activation timer for opening a dialog immediately on page render based on a ms timer. Lastly, other plugins and consumers of this block can open/close or close all dialogs utilizing the Interactivity API, like so: `store('core/dialog').state.dialogs.[{the dialog's elm id}].isOpen = true` or of course identifying all dialogs on page via `store('core/dialog').state.dialogs`.

**Comment by sethrubenstein** [comment]
Also, as I said `core/dialog` was built for maximum extensibility. In addition to its core feature set we've added VideoPress and Youtube support to our block library, snippets here: https://gist.github.com/sethrubenstein/63befb34c78e763b22bcc7a1d9613e00

**Comment by Dionnie** [comment]
One of the reason why Gutenberg is still trash in 2026, you need to be a hardcore nerd wordpress senior developer just to make a popup, god help us all
